presto-base-arrow-flight/src/main/java/com/facebook/plugin/arrow/AbstractArrowMetadata.java

```diff
@@ -0,0 +1,353 @@
+    {
+        switch (intType.getBitWidth()) {
+            case 64:
+                return new ColumnMetadata(columnName, BigintType.BIGINT);
+            case 32:
+                return new ColumnMetadata(columnName, IntegerType.INTEGER);
+            case 16:
+                return new ColumnMetadata(columnName, SmallintType.SMALLINT);
+            case 8:
+                return new ColumnMetadata(columnName, TinyintType.TINYINT);
+            default:
+                throw new ArrowException(ARROW_FLIGHT_ERROR, "Invalid bit width " + intType.getBitWidth());
+        }
+    }
+
+    private ColumnMetadata createFloatingPointColumnMetadata(String columnName, ArrowType.FloatingPoint floatingPointType)
+    {
+        switch (floatingPointType.getPrecision()) {
+            case SINGLE:
+                return new ColumnMetadata(columnName, RealType.REAL);
+            case DOUBLE:
+                return new ColumnMetadata(columnName, DoubleType.DOUBLE);
+            default:
+                throw new ArrowException(ARROW_FLIGHT_ERROR, "Invalid floating point precision " + floatingPointType.getPrecision());
+        }
+    }
+
+    /**
+     * Provides the field type, which can be overridden by concrete implementations
+     * with their own custom type.
+     *
+     * @return the field type
+     */
+    protected Type overrideFieldType(Field field, Type type)
+    {
+        return type;
+    }
+
+    protected abstract ArrowFlightRequest getArrowFlightRequest(ArrowFlightConfig config, Optional<String> query, String schema, String table);
+
+    protected abstract ArrowFlightRequest getArrowFlightRequest(ArrowFlightConfig config, String schema);
+
+    protected abstract String getDataSourceSpecificSchemaName(ArrowFlightConfig config, String schemaName);
+
+    protected abstract String getDataSourceSpecificTableName(ArrowFlightConfig config, String tableName);
+
+    @Override
+    public ConnectorTableHandle getTableHandle(ConnectorSession session, SchemaTableName tableName)
+    {
+        if (!listSchemaNames(session).contains(tableName.getSchemaName())) {
+            return null;
+        }
+
+        if (!listTables(session, Optional.ofNullable(tableName.getSchemaName())).contains(tableName)) {
+            return null;
+        }
+        return new ArrowTableHandle(tableName.getSchemaName(), tableName.getTableName());
+    }
+
+    public List<Field> getColumnsList(String schema, String table, ConnectorSession connectorSession)
+    {
+        try {
+            String dataSourceSpecificSchemaName = getDataSourceSpecificSchemaName(config, schema);
+            String dataSourceSpecificTableName = getDataSourceSpecificTableName(config, table);
+            ArrowFlightRequest request = getArrowFlightRequest(clientHandler.getConfig(), Optional.empty(),
+                    dataSourceSpecificSchemaName, dataSourceSpecificTableName);
+
+            FlightInfo flightInfo = clientHandler.getFlightInfo(request, connectorSession);
```

Comment:
You might want to call `clientHandler.getSchema` directly instead. The default implementation will call `getFlightInfo` but the service could implement `getSchema` to not retrieve the full `FlightInfo`

Reply:
Updated to call `clientHandler.getSchema`